### PR TITLE
[GEOS-11160] GitHub workflow updates for consistency

### PR DESCRIPTION
**Pull request by jodygarnett**
[GEOS-11160](https://osgeo-org.atlassian.net/browse/GEOS-11160) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

In troubleshooting build differences against build.geosever.org I got frustrated with the inconsistency between our workflow actions.

Making list of changes here:

* [x] added assembly.yml to test packaging (thus no longer required in linux_jdk11 and linux_jdk17
* [x] actions/setup-java@v2 --> actions/setup-java@v3
* [x] Use ``pip install -r requirements.txt`` to manage python environment for build docs under version control
  (this replies use of setting up python with packages `sphinx` and `requests` in the workflow)
* [x] checkout@v3 --> checkout@v4, defaulting to depth 0 
* [x] fix names (javdoc.yml was called Linux JDK 11 GitHub CI)
* [x] workflows that precompile without tests use the same mvn command 
* [x] For workflows that precompile without tests, use of `org.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn` to suppress download messages
* [x] consistent ordering of mvn commands with goals at the end of the line
* [x] use matrix and job options to have a single linux.yml workflow testing: 11 and 17
* [x] use experimental matrix parameter to flag JDK 21 as continue-on-error to avoid failing entire workflow
 
This pull request:

- [x] breaks out assembly as a distinct check
- [x] uses `MAVEN_OPTS` in all jobs
- [x] uses `pip install -r doc/en/requirements.txt --update` to configure sphinx-build environment
- [x] consistent ordering of mvn options ending with the maven goals at the end of the line
- [x] use of -B -ntp and  MAVEN_OPTS to reduce the chatter about downloading dependencies
- [x] "quick" builds for javadocs, assembly, etc... do not require fail at end
- [x] steps named consistently
- [x] actions named consistently with "GitHub CI" 
- [x] reduced the checkout depth to the default `1` (why was `500` a good idea?)
- [ ] linux 11 jobs is running out of heap space when building war (it is the only job with -Xmx512m --> 756m, rather than the default of 1/4 RAM)

Experiments not included in this PR:
- I am confused about when and why `-U` and `-nsu` are used (and sometimes on the same line!).
- Tried clearing the maven cache of SNAPSHOTs before build (to avoid use of `-U` but it did not work out
- Tried the maven built into the action/java and it was 3.8.x, the existing maven action provides us greater control of the specific maven version, but greater complexity in managing of maven cache. (Should we simplify?
- The  linux java 11 and java 17 could be handles as a matrix (any reason why not?)
  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [ ] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by jodygarnett** [comment]
When trying to clear the m2 cache (using the delete snippet) before build I was hoping to optimize to avoid the need for -U, which was being repeated for both the -Prelease and the -PcommunityRelease steps.

When I did this:
```
Error: 0,162 [ERROR] Failed to execute goal on project gs-sec-oauth2-google: Could not resolve dependencies for project org.geoserver.community:gs-sec-oauth2-google:jar:2.25-SNAPSHOT: Could not find artifact org.geoserver.community.security:gs-sec-oauth2-web:jar:tests:2.25-SNAPSHOT in osgeo-snapshots (https://repo.osgeo.org/repository/snapshot/) -> [Help 1]
Error: 0,162 [ERROR] Failed to execute goal on project gs-sec-oauth2-github: Could not resolve dependencies for project org.geoserver.community:gs-sec-oauth2-github:jar:2.25-SNAPSHOT: Could not find artifact org.geoserver.community.security:gs-sec-oauth2-web:jar:tests:2.25-SNAPSHOT in osgeo-snapshots (https://repo.osgeo.org/repository/snapshot/) -> [Help 1]
Error: 0,163 [ERROR] Failed to execute goal on project gs-sec-oauth2-geonode: Could not resolve dependencies for project org.geoserver.community:gs-sec-oauth2-geonode:jar:2.25-SNAPSHOT: org.geoserver.community.security:gs-sec-oauth2-web:jar:tests:2.25-SNAPSHOT was not found in https://repo.osgeo.org/repository/snapshot/ during a previous attempt. This failure was cached in the local repository and resolution is not reattempted until the update interval of osgeo-snapshots has elapsed or updates are forced -> [Help 1]
Error: 0,163 [ERROR] Failed to execute goal on project gs-sec-oauth2-openid-connect-core: Could not resolve dependencies for project org.geoserver.community.security:gs-sec-oauth2-openid-connect-core:jar:2.25-SNAPSHOT: org.geoserver.community.security:gs-sec-oauth2-web:jar:tests:2.25-SNAPSHOT was not found in https://repo.osgeo.org/repository/snapshot/ during a previous attempt. This failure was cached in the local repository and resolution is not reattempted until the update interval of osgeo-snapshots has elapsed or updates are forced -> [Help 1]
Error: 0,164 [ERROR] Failed to execute goal on project gs-sec-keycloak: Could not resolve dependencies for project org.geoserver.community:gs-sec-keycloak:jar:2.25-SNAPSHOT: org.geoserver.community.security:gs-sec-oauth2-web:jar:tests:2.25-SNAPSHOT was not found in https://repo.osgeo.org/repository/snapshot/ during a previous attempt. This failure was cached in the local repository and resolution is not reattempted until the update interval of osgeo-snapshots has elapsed or updates are forced -> [Help 1]
```

It appeared that it was not managing to find a SNAPSHOT that should be produced as part of the current run.

**Comment by jodygarnett** [comment]
This works in providing a maven environment c
```
- uses: actions/setup-java@v3
  with:
    distribution: 'temurin'
    java-version: '11'
    cache: 'maven'
```

**Comment by jodygarnett** [comment]
linux MAVEN_OPTS providing more feedback (use with -fae)
```
-Dmaven.wagon.httpconnectionManager.ttlSeconds=25
-Dmaven.wagon.http.retryHandler.count=3
-Xmx756m
-Dorg.slf4j.simpleLogger.showDateTime=true
-Dorg.slf4j.simpleLogger.dateTimeFormat=HH:mm:ss,SSS
```

Quick MAVEN_OPTS for building without tests:
```
-Dmaven.wagon.httpconnectionManager.ttlSeconds=25
-Dmaven.wagon.http.retryHandler.count=3
-Dorg.slf4j.simpleLogger.showDateTime=true
-Dorg.slf4j.simpleLogger.dateTimeFormat=HH:mm:ss,SSS
-Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn
```

**Comment by aaime** [comment]
Sigh, too many changes, each file got significant amounts. Don't have time to go through over them with a fine tooth comb. Turning down review request for now, will likely find time after FOSS4G-NA, though not necessarily the week right after it.

**Comment by jodygarnett** [comment]
> Sigh, too many changes, each file got significant amounts. Don't have time to go through over them with a fine tooth comb. Turning down review request for now, will likely find time after FOSS4G-NA, though not necessarily the week right after it.

Understood, it is the same changes in each file (ie consistency) so I can be sure they are on the same page. I cannot find a good way to do re-use in GitHub workflows.

**Comment by jodygarnett** [comment]
Commenting out JDK 21 experimental as it is a waste of resources until testing is addressed:
```
2023-10-17T05:01:00.2195682Z 05:01:00,219 [ERROR] testCompactHostname(org.geoserver.GeoServerNodeDataTest)  Time elapsed: 0.073 s  <<< ERROR!
2023-10-17T05:01:00.2197594Z org.easymock.cglib.core.CodeGenerationException: java.lang.reflect.InvocationTargetException-->null
2023-10-17T05:01:00.2199264Z 	at org.geoserver.GeoServerNodeDataTest.mockAddress(GeoServerNodeDataTest.java:21)
2023-10-17T05:01:00.2202533Z Caused by: java.lang.reflect.InvocationTargetException
2023-10-17T05:01:00.2203699Z 	at org.geoserver.GeoServerNodeDataTest.mockAddress(GeoServerNodeDataTest.java:21)
2023-10-17T05:01:00.2205594Z Caused by: java.lang.IncompatibleClassChangeError: class $java.net.InetAddress$$EnhancerByCGLIB$$75566629 cannot inherit from sealed class java.net.InetAddress
```

**Comment by aaime** [comment]
Looks like we'll have to migrate away from cglib in order to support Java 21: https://github.com/cglib/cglib (see warning at the top of the readme)

**Comment by groldan** [comment]
FWIW I've started a branch to get the build working with Java 21, see #7207
So far only core modules, but there was only one test Easymock couldn't handle, and it wasn't even needed anyway.

File: .github/workflows/linux.yml [hunk 1]
```diff
@@ -11,42 +11,49 @@ env:
...
     - name: Build with Maven
-      run: mvn -B -U -T3 -fae -Dspotless.apply.skip=true -Prelease -f src/pom.xml clean install
-    - name: Build community modules
-      run: mvn -nsu -B -U -T4 -fae -Dspotless.apply.skip=true -DskipTests -Prelease -PcommunityRelease -f src/community/pom.xml clean install
-    - name: Package
-      run: mvn -f src/pom.xml assembly:single -nsu -N
-    - name: Package community modules
-      run:  mvn -f src/community/pom.xml assembly:single -nsu -N
+      run: |
+        mvn -B -ntp -U -T3 -fae -Dspotless.apply.skip=true -Prelease -f src/pom.xml clean install
```
[the remark this answers is not in the document]
Comment (aaime) [reply .github/workflows/linux.yml:53]: Also, and mind, just thinking out loud, there is quite a bit of repetition with the maven opts repeated over and over. Looks like there is a way to centralize them:

https://docs.github.com/en/actions/using-workflows/reusing-workflows

That said, just thinking out loud, not a request, just an idea.

File: .github/workflows/linux_jdk17.yml [hunk 2]
```diff
@@ -1,53 +0,0 @@
-name: Linux JDK 17 GitHub CI
```
Comment (aaime) [inline .github/workflows/linux_jdk17.yml:1]: The original reason to have a separate workflow was to allow restarting it in isolation... when we did it, all the workflows in the same action had to be restarted, if one failed. Not sure if that is no longer the case.... [pheraps not](https://stackoverflow.com/questions/71315332/only-re-run-the-failed-job-in-a-github-actions-matrix).

File: .github/workflows/linux.yml [hunk 3]
```diff
@@ -11,42 +11,49 @@ env:
...
+      uses: actions/setup-java@v3
       with:
         java-version: ${{ matrix.jdk }}
         distribution: ${{ matrix.dist }}
     - name: Set up Maven
-      uses: stCarolas/setup-maven@v4
+      uses: stCarolas/setup-maven@v4.5
       with:
-        maven-version: 3.8.4
+        maven-version: 3.9.5
```
[the remark this answers is not in the document]
Comment (mprins) [reply .github/workflows/linux.yml:43]: GH pretty much stays up2date with the Maven version nowadays. Currently 3.8.8, see https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md
I tend to rely on that unless developing Maven plugins

File: .github/workflows/linux_jdk17.yml (hunk 2 above)
Comment (mprins) [reply .github/workflows/linux_jdk17.yml:1]: Matrix jobs have cone a long way, it is now possible to restart a single job from a workflow

File: .github/workflows/assembly.yml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    - name: Build GeoServer modules and extensions without tests
+      run: |
+        mvn --version
+        mvn -B -ntp -U -T3 -DskipTests -Dspotless.apply.skip=true -Prelease -f src/pom.xml install
+    - name: Package GeoServer modules and extensions
+      run: |
+        mvn -B -ntp -nsu -N -f src/pom.xml assembly:single
+    - name: Build community modules without tests
+      run: |
+        mvn -B -ntp -U -T3 -DskipTests -Dspotless.apply.skip=true -Prelease,communityRelease -f src/community/pom.xml install
```
[the remark this answers is not in the document]
Comment (jodygarnett) [reply .github/workflows/assembly.yml:45]: Probably not! I will remove it …

File: .github/workflows/linux.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (jodygarnett) [reply .github/workflows/linux.yml:43]: When I left off the version number it ended up with an older version of maven. Burnie we could leave it off and stay with whatever GitHub has configured.

File: .github/workflows/linux.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (jodygarnett) [reply .github/workflows/linux.yml:53]: We can add that to maven_opts for the non QA workflows. 

I had not looked into reusable workflows yet. For is the main reusable workflow is quickly building in order to run a specific test.

File: .github/workflows/linux.yml (hunk 3 above)
Comment (mprins) [inline .github/workflows/linux.yml:43]: You'd remove the whole `setup-maven` build step:

```yaml
      uses: stCarolas/setup-maven@v4.5
      with:
        maven-version: 3.9.5
```

**Comment by jodygarnett** [comment]
Updated based on feedback, -Dspotless.apply.skip=true included in MAVEN_OPTS.
Also doubled checked the order of the mvn commands again.

**Comment by jodygarnett** [comment]
All feedback has addressed, bumping up to -Xmx1024m as I am tried of restarting builds for maven-war-plugin heapspace failures.

Andrea: I have an experiment with workflow reuse, but it is not so promising yet. It can reduce code duplication but at great cost to complexity (trying to pass a checkout and build from one job to another as a "cache"). I will share the experiment as a distinct PR and ask if it is worth pursuing.

**Pull request merged**